internal/worker/events.go

```diff
@@ -0,0 +1,172 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/webhooks"
+	"go.uber.org/zap"
+)
+
+type (
+	EventManager interface {
+		AddSubscriber(id string, s EventSubscriber) (chan struct{}, error)
+		HandleEvent(event webhooks.Event)
+		Run(ctx context.Context, eventURL string, opts ...webhooks.HeaderOption) error
+		Shutdown(context.Context) error
+	}
+
+	EventSubscriber interface {
+		HandleEvent(event webhooks.Event) error
+		Subscribe(e EventManager) error
+	}
+
+	WebhookManager interface {
+		RegisterWebhook(ctx context.Context, wh webhooks.Webhook) error
+		UnregisterWebhook(ctx context.Context, wh webhooks.Webhook) error
+	}
+)
+
+type (
+	eventManager struct {
+		webhooks WebhookManager
+		logger   *zap.SugaredLogger
+
+		registerInterval time.Duration
+
+		mu         sync.Mutex
+		subs       map[string]eventSubscriber
+		registered []webhooks.Webhook
+	}
+
+	eventSubscriber struct {
+		EventSubscriber
+		readyChan chan struct{}
+	}
+)
+
+func NewEventManager(w WebhookManager, l *zap.Logger, registerInterval time.Duration) EventManager {
+	return &eventManager{
+		webhooks: w,
+		logger:   l.Sugar().Named("events"),
+
+		registerInterval: registerInterval,
+
+		subs: make(map[string]eventSubscriber),
+	}
+}
+
+func (e *eventManager) AddSubscriber(id string, s EventSubscriber) (chan struct{}, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	_, ok := e.subs[id]
+	if ok {
+		return nil, fmt.Errorf("subscriber with id %v already exists", id)
+	}
+
+	readyChan := make(chan struct{})
```

Comment:
Why not create a single `readyChan` in `NewEventManager` and return that? Seems like once the manager is registered all channels get closed so a single channel should work as well.

Then you also don't need the `eventSubscriber` type and the `subs` map can be `map[string]EventSubscriber` or even just `[]EventSubscriber` since it doesn't seem like subscribers ever want to unsubscribe.